### DISCUSSION TITLE
Do I need to pad my training images?

### DISCUSSION THREAD

**Discussion by Paasu** [neuropoly/axondeepseg#295] [Help needed] [answered]
I have been training my images recently but have been running into issues, I have padded my images to make sure they are big enough for segmentation but does the padding itself change the training results? The images were originally 4008x2672 pixels but then I padded it to 6000x4000 pixels.

**Comment by vasudev-sharma** [comment]
Hey @Paasu, could you post the log as it would enable to debug the problem ?

**Comment by mathieuboudreau** [comment, marked as answer]
Hi @Paasu,

No, you shouldn't need to pad your training images. If your images are too small to create the patch size that you desire, this would typically be because the ratio of your actual pixel size of your images (determined by your microscope) over the resampling pixel size (used for creating uniform resolutions in images during training) isn't in the correct range to support the patch size you need (e.g. 512x512). 

To fix this, you can try decreasing or increasing the value of the parameter `general_pixel_size` in the [dataset building notebook](https://github.com/neuropoly/axondeepseg/blob/master/notebooks/01-guide_dataset_building.ipynb). To have an idea of the range in values you would need, take the smallest dimension of your image (e.g. 2672), multiply it by your actual image resolution, then try dividing this number by the `general_pixel_size` value. For example, if your actual image resolution is 0.001 um, then a ressampling resolution of 0.1 would be too large as can be seen from that operation:  2672\*0.01/0.1 = 267.2 < 512 (patch size). In this case, a `general_pixel_size` value of 0.02 might be within the good range such that your image won't need padding: 2672\*0.01/0.02 = 1336 > patch size of 512. 

The smaller the `general_pixel_size` value, the more patches created, and thus the less information about the axons will be available in each patches, so you want to find a balance here. Typically, 6 to 20 patches per images would be a good range to aim for in the beginning.

**Comment by Paasu** [comment]
My actual image resolution is 0.5 microns per 300 pixels, so the value would be 0.00166666666 um. The value I get following your operation is 0.0086 under 512 patch size. This generates around 3 patches per image, so I decreased it to 256 patch size and that gives me 11 patches per image so I went with that. Will this make a strong difference? And is it possible to video call in anyway? I have already sent you an email with some additional details.

**Comment by mathieuboudreau** [comment]
Hi @Paasu,

Yes that setting sounds like it could work, but it also depends on the size of your axons relative to your patches. If you take a look at the patched images generated as an output to the data preparation notebook, are axons mostly fully contained within each patch, or are they cut off such that each patch only has a corner of an axon within it?

We also had better results using 512x512 patches, but maybe @alzaia can expand on his experience with 256x512 patches (where these more computationally intensive, or took longer to train?)

My apologies about your email, I can't seem to find it in my mailbox. Would you mind resending it? My email is in my user profile: https://github.com/mathieuboudreau. We'd be happy to have a video call with you.

**Comment by Paasu** [comment]
![Figure_2](https://user-images.githubusercontent.com/50454820/82564080-bfa18280-9baa-11ea-9678-f682187014ee.png)
Above is the figure image spewed out after 600 epochs.
![image_10](https://user-images.githubusercontent.com/50454820/82564123-cfb96200-9baa-11ea-8082-5678f2074a84.png)
Here is one of the patches created in the patchedImages folder. It looks like at times the axons are in fact, cut off. As shown above but also here
![mask_0](https://user-images.githubusercontent.com/50454820/82564203-f1b2e480-9baa-11ea-85b8-b7903410537b.png)
and here
![mask_4](https://user-images.githubusercontent.com/50454820/82564218-f8d9f280-9baa-11ea-9694-9ed114bc2299.png)
I have also sent you another email regarding the video call
